### PR TITLE
feat: adding support for aws list_objects_v2

### PR DESCRIPTION
**Pull request by slawrence87544**
body: In order to get the inventory of observation files stored in
our BDP aws s3 bucket 'noaa-reanalyses-pds' we needed to add
support for the boto3 function list_objects_v2.  We also needed
to add handling for the database insertion for the command result
and the individual file meta.

### CODE REVIEW

File: src/obs_inv_utils/aws_s3_interface.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        if m is not None and m.group(0) is not None:
+            print('Only a-z A-Z 0-9 and - . / _ characters allowed in filepath')
+            raise ValueError(f'Invalid characters found in file path: {arg}')
+    except Exception as e:
+        raise ValueError(f'Invalid file path: {e}')
+
+    return True
+
+
+def s3_object_list_v2_parser(obj_list_contents, obs_cycle_time):
```
Comment (frolovsa) [inline src/obs_inv_utils/aws_s3_interface.py:115]: why there is a "v2" string in the name. does it mean version 2?

File: src/obs_inv_utils/search_engine.py [hunk 2]
```diff
@@ -176,6 +194,46 @@ def parse_filename(filename):
     return filename_meta
 
 
+def process_aws_s3_list_objects_v2_resp(contents):
```
Comment (frolovsa) [inline src/obs_inv_utils/search_engine.py:197]: same question about "v2"

File: src/obs_inv_utils/aws_s3_interface.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        'arg_validator',
+        'output_parser'
+    ],
+)
+
+AWS_BDP_BUCKET = 'noaa-reanalyses-pds'
+CMD_GET_S3_OBJ_LIST = 'list_objects'
+
+
+def get_bdp_s3_client():
```
Comment (frolovsa) [inline src/obs_inv_utils/aws_s3_interface.py:71]: will your code also work for a private bucket or just for bdp?

File: src/obs_inv_utils/search_engine.py (hunk 2 above)
Comment (frolovsa) [reply src/obs_inv_utils/search_engine.py:197]: This is a suggestion: Should hpss and aws specific functionality reside in hpss and aws specific modules and the search engine should be oblivious if hpss or aws is used?

File: src/obs_inv_utils/search_engine.py [hunk 4]
```diff
@@ -215,6 +273,31 @@ def process_inspect_tarball_resp(contents):
     tbl_factory.insert_obs_inv_items(tarball_files_meta)
 
 
+def default_datetime_converter(obj):
+   if isinstance(obj, datetime):
+      return obj.__str__()
+
+def post_aws_s3_cmd_result(raw_response, obs_cycle_time):
```
Comment (frolovsa) [inline src/obs_inv_utils/search_engine.py:280]: same here. should this be inside of aws_s3_interface

File: src/obs_inv_utils/aws_s3_interface.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
+import re
```
Comment (frolovsa) [inline src/obs_inv_utils/aws_s3_interface.py:1]: suggestion on the naming of this file "src/obs_inv_utils/aws_s3_interface.py"
should we name all interfaces like "interface_aws_s3", "interface_hpss"

**Review by frolovsa — CHANGES_REQUESTED** [review]
Steve, thank you for this addition. I didn't dive into the details f the implementation. I do have a few questions/suggestions about function names, file naming convention. I also wonder if the aws and hpss specifications should be localized fully in the interface modules and search_engine should be completely oblivious to the implementation. 

That said, we should find a good balance between the speed of the development and how clean we want this code to be with the abstraction etc.

File: src/obs_inv_utils/aws_s3_interface.py (hunk 5 above)
Comment (slawrence87544) [reply src/obs_inv_utils/aws_s3_interface.py:1]: Sure, I can do that.

File: src/obs_inv_utils/aws_s3_interface.py (hunk 3 above)
Comment (slawrence87544) [reply src/obs_inv_utils/aws_s3_interface.py:71]: For now, it only works for the BDP bucket, but can easily be modified to allow private buckets.

File: src/obs_inv_utils/aws_s3_interface.py (hunk 1 above)
Comment (slawrence87544) [reply src/obs_inv_utils/aws_s3_interface.py:115]: That is a carryover from the boto3 function definition.  I can leave it out if it seems confusing.

File: src/obs_inv_utils/search_engine.py (hunk 2 above)
Comment (slawrence87544) [reply src/obs_inv_utils/search_engine.py:197]: That is my goal, yes.  I just wanted to produce something quickly to scan/spider the aws inventory.  I can create a new pull request (dependent on this one) which makes search engine agnostic.

File: src/obs_inv_utils/search_engine.py (hunk 4 above)
Comment (slawrence87544) [reply src/obs_inv_utils/search_engine.py:280]: If you prefer, I can include the changes necessary to separate out the interface specificity in this PR, but it does make the PR bigger.

File: src/obs_inv_utils/aws_s3_interface.py (hunk 1 above)
Comment (frolovsa) [reply src/obs_inv_utils/aws_s3_interface.py:115]: if this is inherited from boto then i am ok keeping it

File: src/obs_inv_utils/search_engine.py (hunk 2 above)
Comment (frolovsa) [reply src/obs_inv_utils/search_engine.py:197]: Maybe make an issue and address it in a separate PR after this one is merged?

File: src/obs_inv_utils/search_engine.py (hunk 4 above)
Comment (frolovsa) [reply src/obs_inv_utils/search_engine.py:280]: same here. follow up PR will be better

**Review by frolovsa — APPROVED** [review]

**Comment by HenryRWinterbottom** [comment]
Looks fine to me.

**Review by HenryRWinterbottom — APPROVED** [review]

File: src/obs_inv_utils/aws_s3_interface.py (hunk 1 above)
Comment (slawrence87544) [reply src/obs_inv_utils/aws_s3_interface.py:115]: Yes, it is directly inherited from boto3.

File: src/obs_inv_utils/search_engine.py (hunk 4 above)
Comment (slawrence87544) [reply src/obs_inv_utils/search_engine.py:280]: I'll create a follow-up PR for this one.

File: src/obs_inv_utils/aws_s3_interface.py (hunk 5 above)
Comment (slawrence87544) [reply src/obs_inv_utils/aws_s3_interface.py:1]: After looking at this again, I remembered why I chose that naming convention.  It follows the conventions set by UFS-RNR.  Namely, [specificity]_[method_type].py  I'd like to keep this convention if it's ok with the group.

**Review by frolovsa — APPROVED** [review]

**Pull request merged**